### PR TITLE
Pass stuff in to XFEM through the parameters

### PR DESCRIPTION
**Pull request by bwspenc**
ref #6446

### CODE REVIEW

File: framework/include/base/XFEMInterface.h [hunk 1]
```diff
@@ -48,9 +49,9 @@ class XFEMInterface : public ConsoleStreamInterface
    * Constructor
    */
   explicit
-  XFEMInterface(MooseApp & app, const MooseSharedPointer<FEProblem> fe_problem):
-    ConsoleStreamInterface(app),
-    _fe_problem(fe_problem),
+  XFEMInterface(const InputParameters & params) :
+    ConsoleStreamInterface(*params.get<MooseApp *>("_moose_app")),
```
Comment (permcody) [inline framework/include/base/XFEMInterface.h]: Use this method instead:
https://github.com/idaholab/moose/blob/devel/framework/include/utils/InputParameters.h#L112

Same for the line of code below

**Comment by permcody** [comment]
Thanks! I've activated the "Heavy Valgrind Modules" target. Let's see if this helps.

File: modules/xfem/src/actions/XFEMAction.C [hunk 2]
```diff
@@ -62,7 +62,8 @@ XFEMAction::act()
   MooseSharedPointer<XFEMInterface> xfem_interface = _problem->getXFEM();
   if (xfem_interface == NULL)
   {
-    MooseSharedPointer<XFEM> new_xfem (new XFEM(_problem->getMooseApp(), _problem));
+    _pars.set<FEProblem *>("_fe_problem") = &*_problem;
```
Comment (bwspenc) [inline modules/xfem/src/actions/XFEMAction.C]: @permcody what do you think of this? I'm getting a raw pointer to FEProblem from the reference to the shared pointer that we have in Action. I know that this defeats the smartness of the smart pointer, but there are raw pointers to FEProblem all over the place in MOOSE, and there's no ownership of that pointer in XFEM. Based on the recent observation by @jiangwen84, switching this over to a raw pointer probably fixes all of those valgrind errors. I'm building this on Linux right now to run with valgrind to see if that is the case.

Comment (permcody) [inline modules/xfem/src/actions/XFEMAction.C]: I think this is fine for now. We'll be able to clean some of this up with C++11. There's a "weak pointer" concept to handle these "ring reference" situations. If this fixes the issue, I'll be happy.

**Comment by bwspenc** [comment]
@permcody I guess you'll see the results of the valgrind testing eventually, but I just ran them myself with valgrind on the xfem module and confirmed that this fixes all of the memory leaks we were seeing in the xfem module tests.

**Comment by permcody** [comment]
Excellent!

**Comment by jiangwen84** [comment]
👍 💯

**Pull request merged**